problem/0000-metrics-v1.md

```diff
@@ -0,0 +1,136 @@
+# Linkerd Metrics v1
+
+- Contribution Name: metrics-v1
+- Implementation Owner:
+- Start Date: 2020-04-23
+- Target Date:
+- RFC PR:
+- Linkerd Issue:
+- Reviewers: @grampelberg @adleong
+
+## Problem Statement
+
+[problem-statement]: #problem-statement
+
+Linkerd's `stat` API powers both the command-line and web interfaces, but it's tailored closely
+to their use cases and it has evolved only to express the features of these tools. Because of its
+narrow focus, integrators tend to read data directly from Prometheus, which features an
+expressive query language and a variety of external integrations/extensions. This presents an
+implicit, undocumented relationship between the proxy's metrics exposition structure and traffic
+in Kubernetes.
+
+The goal of this RFC is to define an external-facing API for Linkerd that:
+
+1. The semantics of which are defined by this RFC;
+2. Is stable/versioned;
+3. Supports Linkerd's UIs as well as arbitrary external clients;
+4. Supports Kubernetes RBAC;
+5. Supports extensions for additional protocols, metrics, and metadata; and
+6. Supports arbitrary workload hierarchies (i.e for custom operators).
+
+## Design Notes
+
+### API
+
+The metrics API is read-only.
+
+In order to support RBAC, all requests should be scoped to the resources on which metrics are
+measured.
+
+It is time-oriented. It should have both time-series and summary capabilities.
```

Comment:
How much history is available from this API?  How can users extend the amount of history available?

Today, users can federate Linkerd's Prometheus into their own Prometheus with a longer retention period.  Is it a goal of this RFC that users would no longer federate directly from Linkerd's Prometheus?  Does this API support the use-case of federating Linkerd's metrics into longer term storage?

Reply:
At an API level there are no restrictions. Operators are able to tune/configure this to their needs. The API returns whatever data is available in the underlying storage.  We will need to publish docs/config/tools/etc for Linkerd scrape configs that provided the metrics needed to power the API (though this is out of scope for this RFC, imo... this should focus just at the API-level)

Reply:
Users today who want a retention period of longer than 6 hours federate the Linkerd prometheus into their own.  If the plan is to declare that Linkerd's Prometheus is not part of its API then I think we need to figure out how to support this use-case.  

e.g. Should the API support bulk extraction of metrics into an external longer-term storage?  Or should Linkerd API surface area include knobs for increasing the internal retention period?